src/upgrades/v0_9_0/pubd_objects_migration.rs

```diff
@@ -0,0 +1,506 @@
+
+                    self.archive_migrated(&event_key)?;
+
+                    if old_evt.needs_migration() {
+                        info.last_event += 1;
+
+                        events.push(info.last_event);
+                        let migrated_event = old_evt.into_stored_pubd_event(info.last_event)?;
+                        let key = KeyStoreKey::scoped(scope.to_string(), format!("delta-{}.json", info.last_event));
+                        self.store.store(&key, &migrated_event)?;
+                    }
+                }
+
+                if events.is_empty() {
+                    continue; // This command has no relevant events in 0.9, so don't save it.
+                }
+
+                old_cmd.effect = OldStoredEffect::Events(events);
+            }
+
+            old_cmd.version = info.last_event + 1;
+            old_cmd.sequence = info.last_command;
+
+            info.last_command += 1;
+            info.last_update = old_cmd.time;
+
+            let migrated_cmd = old_cmd.into_pubd_command();
+            let cmd_key = CommandKey::for_stored(&migrated_cmd);
+            let key = KeyStoreKey::scoped(scope.to_string(), format!("{}.json", cmd_key));
+
+            self.store.store(&key, &migrated_cmd)?;
+        }
+
+        // move out the snapshots, we will rebuild from events
+        // there will not be too many now that the publication
+        // deltas are no longer done as events
+        self.archive_snapshots(&scope)?;
+
+        // update the info file
+        info.snapshot_version = 0;
+        info.last_command -= 1;
+        self.store.store(&info_key, &info)?;
+
+        // verify that we can now rebuild the 0.9 publication server based on
+        // migrated commands and events.
+        self.new_store.warm().map_err(|e| UpgradeError::Custom(format!("Could not rebuild state after migrating pubd! Error was: {}. Please report this issue to rpki-team@nlnetlabs.nl. For the time being: restore all files in the 'migration-0.9' directory to their parent directory and revert to the previous version of Krill.", e)))?;
+
+        // Great, we have migrated everything, now delete the archived
+        // commands and events which are no longer relevant
+        self.drop_migration_scope(scope)?;
+
+        Ok(())
+    }
+
+    fn store(&self) -> &KeyValueStore {
+        &self.store
+    }
+
+    fn version_before(kv: &KeyValueStore, before: KeyStoreVersion) -> Result<bool, UpgradeError> {
+        let key = KeyStoreKey::simple("version".to_string());
+        match kv.get::<KeyStoreVersion>(&key) {
+            Err(e) => Err(UpgradeError::KeyStoreError(e)),
+            Ok(None) => Ok(true),
+            Ok(Some(current_version)) => Ok(current_version < before),
+        }
+    }
+}
+
+/// Pre 0.9 Repository which combines the access (ID) functions,
+/// and content. This is split up 0.9 into two separate components.
```

Comment:
This is split up *in* 0.9 into two separate components?

Reply:
fixed comment (I hope)